File: Problems/0020_Valid_Parentheses/0020_Valid_Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    var isValidHelper = function(s,prevBracket){//helper for previous
        if(s.length>0){
            var curBracket = s.charAt(0);
            console.log(curBracket);
            if(curBracket==")"){
                if(prevBracket!="("){return false;}
                else{return isValidHelper(s.substring(1,s.length-1));}
            }
            else if(curBracket=="}"){
                if(prevBracket!="{"){return false;}
                else{return isValidHelper(s.substring(1,s.length-1));}
            }
            else if(curBracket=="]"){
                if(prevBracket!="["){return false;}
                else{return isValidHelper(s.substring(1,s.length-1));}
            }
            else{
                return isValidHelper(s.substring(1,s.length-1),curBracket);
            }
            
        }
        return true;
        //if s is empty, return nothing
    }
    return isValidHelper(s,"");
};

File: Problems/0020_Valid_Parentheses/0020_Valid_Parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> sStack = new Stack<Character>();
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)==')'){
                char c=sStack.empty()?'#':sStack.pop();
                if(c!='(') return false;
            }
            else if(s.charAt(i)=='}'){
                char c=sStack.empty()?'#':sStack.pop();
                if(c!='{') return false;
            }
            else if(s.charAt(i)==']'){
                char c=sStack.empty()?'#':sStack.pop();
                if(c!='[') return false;
            }
            else{
                sStack.push(s.charAt(i));
            }
        }
        return sStack.empty();
    }
}